Mathlib/Order/SuccPred/Arithmetic.lean

```diff
@@ -0,0 +1,193 @@
+-/
+
+/-- A typeclass for `succ x = x + 1`. -/
+class SuccAddOrder (α : Type*) [Preorder α] [Add α] [One α] extends SuccOrder α where
+  succ_eq_add_one (x : α) : succ x = x + 1
+
+/-- A typeclass for `pred x = x - 1`. -/
+class PredSubOrder (α : Type*) [Preorder α] [Sub α] [One α] extends PredOrder α where
+  pred_eq_sub_one (x : α) : pred x = x - 1
+
+variable {α : Type*} {x y : α}
+
+namespace Order
+
+section Preorder
+
+variable [Preorder α]
+
+theorem succ_eq_add_one [Add α] [One α] [SuccAddOrder α] (x : α) : succ x = x + 1 :=
+  SuccAddOrder.succ_eq_add_one x
+
+theorem pred_eq_sub_one [Sub α] [One α] [PredSubOrder α] (x : α) : pred x = x - 1 :=
+  PredSubOrder.pred_eq_sub_one x
+
+theorem add_one_le_of_lt [Add α] [One α] [SuccAddOrder α] (h : x < y) : x + 1 ≤ y := by
+  rw [← succ_eq_add_one]
+  exact succ_le_of_lt h
+
+theorem le_sub_one_of_lt [Sub α] [One α] [PredSubOrder α] (h : x < y) : x ≤ y - 1 := by
+  rw [← pred_eq_sub_one]
+  exact le_pred_of_lt h
+
+theorem add_one_le_iff_of_not_isMax [Add α] [One α] [SuccAddOrder α] (hx : ¬ IsMax x) :
+    x + 1 ≤ y ↔ x < y := by
+  rw [← succ_eq_add_one, succ_le_iff_of_not_isMax hx]
+
+theorem le_sub_one_iff_of_not_isMin [Sub α] [One α] [PredSubOrder α] (hy : ¬ IsMin y) :
+    x ≤ y - 1 ↔ x < y := by
+  rw [← pred_eq_sub_one, le_pred_iff_of_not_isMin hy]
+
+theorem add_one_le_iff [Add α] [One α] [SuccAddOrder α] [NoMaxOrder α] : x + 1 ≤ y ↔ x < y :=
+  add_one_le_iff_of_not_isMax (not_isMax x)
+
+theorem le_sub_one_iff [Sub α] [One α] [PredSubOrder α] [NoMinOrder α] : x ≤ y - 1 ↔ x < y :=
+  le_sub_one_iff_of_not_isMin (not_isMin y)
+
+@[simp]
+theorem succ_zero [AddZeroClass α] [One α] [SuccAddOrder α] : succ (0 : α) = 1 := by
+  rw [succ_eq_add_one, zero_add]
+
+@[simp]
+theorem pred_zero [SubNegMonoid α] [One α] [PredSubOrder α] : pred (0 : α) = -1 := by
+  rw [pred_eq_sub_one, zero_sub]
+
+theorem succ_one [AddMonoidWithOne α] [SuccAddOrder α] : succ (1 : α) = 2 := by
+  rw [succ_eq_add_one, one_add_one_eq_two]
+
+@[simp]
+theorem pred_one [AddGroup α] [One α] [PredSubOrder α] : pred (1 : α) = 0 := by
+  rw [pred_eq_sub_one, sub_self]
+
+theorem add_succ [AddSemigroup α] [One α] [SuccAddOrder α] (x y : α) :
+    x + succ y = succ (x + y) := by
+  rw [succ_eq_add_one, succ_eq_add_one, add_assoc]
+
+theorem add_pred [AddGroup α] [One α] [PredSubOrder α] (x y : α) :
+    x + pred y = pred (x + y) := by
+  rw [pred_eq_sub_one, pred_eq_sub_one, add_sub]
+
+theorem succ_add [AddCommSemigroup α] [One α] [SuccAddOrder α] (x y : α) :
+    succ x + y = succ (x + y) := by
+  rw [add_comm, add_succ, add_comm]
+
+theorem pred_add [AddCommGroup α] [One α] [PredSubOrder α] (x y : α) :
+    pred x + y = pred (x + y) := by
+  rw [add_comm, add_pred, add_comm]
+
+theorem natCast_succ [AddMonoidWithOne α] [SuccAddOrder α] (n : ℕ) : n.succ = succ (n : α) := by
+  cases n with
+  | zero => rw [Nat.cast_zero, succ_zero, Nat.cast_one]
+  | succ n => rw [succ_eq_add_one, Nat.cast_add_one, Nat.cast_succ]
```

Comment:
I see this new API as a way to have an algebraic successor/predecessor API. I see no point stating *any* algebraic lemma about `succ`/`pred`! You should instead be stating order theoretic lemmas about `+ 1`/`- 1`.

Reply:
Most of these are generalizations of lemmas that already existed for other types. Is the idea that `x + 1` should be the simp-normal form for `succ x` moving forward?

Reply:
In fact, maybe it might be a good idea to do the ordinal refactor first. That way we can register `succ_eq_add_one` as a `simp` lemma without it causing conflicts.

Reply:
Thought about it. It's definitely better to do this refactor first, since we can then use the new lemmas to rewrite the proofs during the ordinal refactor.

Reply:
I kept `succ_zero` since it was useful in proving other lemmas like `1 ≤ x ↔ 0 < x`.

Reply:
> Is the idea that `x + 1` should be the simp-normal form for `succ x` moving forward?

Yes, at least that's the point of this typeclass